package/MDAnalysis/analysis/MSD.py

```diff
@@ -0,0 +1,468 @@
+        # dictionaries with key: (id) and value: (array index)
+        # selection X frame X dictionary
+        dict_list = [[dict() for j in range(self.len_msd)]
+                     for i in range(n_sel)]
+        # atom ids which satisfy the selection at each frame
+        set_list = [[set() for j in range(self.len_msd)]
+                    for i in range(n_sel)]
+
+        # pre-allocate position array
+        pos = [[np.array([]) for j in range(self.len_msd)]
+               for i in range(n_sel)]
+
+        print("Populating deque...")
+        for ts in self.universe.trajectory[self.t0:self.t0
+                                           + self.len_msd]:
+            this_frame = ts.frame
+            selections = self._select(this_frame)
+
+            for i in range(n_sel):
+                if selections[i] is not None:  # if there is data
+                    pos[i][this_frame-self.t0] = selections[i].positions
+                    
+                    temp_list = selections[i].atoms.ids
+                    # store set of atom ids which satisfy selection
+                    set_list[i][this_frame-self.t0] = set(temp_list)
+                    # link atom id to array index
+                    for j in range(len(temp_list)):
+                        dict_list[i][this_frame-self.t0][temp_list[j]] = j
+                    
+            if this_frame % (self.len_msd/10) == 0:
+                print("Frame: "+str(this_frame))
+        # converting lists to deques
+        for i in range(n_sel):
+            dict_list[i] = deque(dict_list[i], maxlen=self.len_msd)
+            set_list[i] = deque(set_list[i], maxlen=self.len_msd)
+            pos[i] = deque(pos[i], maxlen=self.len_msd)
+        print("Complete!")
+
+        return pos, set_list, dict_list
+
+    def _update_deques(self, pos, set_list, dict_list, this_restart):
+        """Updates lists necessary for MSD calculations
+
+        Parameters:
+        -----------
+        pos : list
+            contains position information, organized:
+            selection X frame X atom X (x, y, z)
+        set_list : list
+            contains set of atoms which satisfy each selection at each
+            frame, organized:
+            selection X frame X set of atom ids
+        dict_list : list
+            contains dictionaries which map atom ids to corresponding
+            pos index, organized:
+            selection X frame X dict
+        this_restart : int
+            frame number of the current restart
+        """
+        # stop updating when there are no more frames to analyze
+        top_frame = this_restart + self.len_msd
+        calc_cutoff = self.tf
+        if self.max_data:
+            calc_cutoff = len(self.universe.trajectory) - 1
+
+        if top_frame+self.dt_restart > calc_cutoff:
+            # feed in zeros
+            for i in range(len(self.select_list)):
+                pos[i].append(np.array([]))
+                set_list[i].append(set())
```

Comment:
This isn't a good variable name. It doesn't me any clue about what is stored in that variable. Instead of stating the type in the name state the content that makes the algorithm easier to understand. I always try to pick variable names that allow me to know what the variable contains just from the name even if I worked on something completely for a year before looking at the code again.